### PR TITLE
added api manager files and react redux thunk.

### PR DESCRIPTION
**Pull request by MuhammadMuneebZafar**

### CODE REVIEW

File: src/redux/store/index.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import { createStore, applyMiddleware } from "redux";
+import thunkMiddleware from "redux-thunk";
+import rootReducer from "../reducers";
+import { composeWithDevTools } from "redux-devtools-extension";
+
+const store = createStore(
+  rootReducer,
+  composeWithDevTools(applyMiddleware(thunkMiddleware)) //composeWithDevTools is used here for development purposes. It should be removed for production
```
Comment (hamza-mubeen-07) [inline src/redux/store/index.ts:8]: i think we should apply check here on `environment` variable that we will get form .env . if production then we will remove `composeWithDevTools` we always forget that we need to remove this for production. what do you think.

File: src/redux/actions/book/bookAction.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import {
+  GET_CATEGORY_DROPDOWN_ITEMS,
+  iGetCategoryOptionsActionType,
+} from "../../types/book/bookType";
+
+export const getCategoryDropdownAction = (
+  categoryDropdownItems: any
```
Comment (hamza-mubeen-07) [inline src/redux/actions/book/bookAction.ts]: kindly change the type `any` to `iCategoryDropdownOption[]`

File: src/routes/routeConstants.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+export const RoutesPath = {
```
Comment (hamza-mubeen-07) [inline src/routes/routeConstants.ts]: what convention will we be following for constants ?  we have upper snake case in recent project.

File: src/services/context/GlobalContext.tsx [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+import { iApiResponse } from "../api-manager/apiManagerInterfaces";
+import { iGetCategoryOptionsActionType } from "../../redux/types/book/bookType";
+import { Dispatch } from "redux";
+
+export interface iGlobalContext {
+  showLoader: boolean;
+  setShowLoader: React.Dispatch<boolean>;
+  globalDispatch: (
+    effect: (
+      dispatch: Dispatch<iGetCategoryOptionsActionType>
```
Comment (hamza-mubeen-07) [inline src/services/context/GlobalContext.tsx]: dispatch will possibly have type of all the actions  so i think i will be something like  `bookActionTypes | userActionTypes`.

File: src/component/home/Home.tsx [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+  const getCategory = async () => {
+    setShowLoader(true);
+    await globalDispatch(getCategoryDropDownEffect());
+  };
+
+  return (
+    <>
+      <h1>Home</h1>
+      {showLoader && <h1>Hello</h1>}
+      <button onClick={() => getCategory()}>Go Error</button>
```
Comment (hamza-mubeen-07) [inline src/component/home/Home.tsx]: kindly replace `() => getCategory()` to `getCategory`.

File: src/redux/effects/effects-for-services/Effect-service.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+import { customDispatch } from "../../../services/api-manager/apiUtilities";
+import { iApiResponse } from "../../../services/api-manager/apiManagerInterfaces";
+
+export const getCategoryDropDownEffect =
+  () =>
+  async (
+    dispatch: Dispatch<iGetCategoryOptionsActionType>
+  ): Promise<iApiResponse> => {
+    const response = await getCategoryService();
+    return customDispatch(
```
Comment (hamza-mubeen-07) [inline src/redux/effects/effects-for-services/Effect-service.ts]: kindly change name of the `customDispatch` it is conflicting with `globalDispatch`.

File: src/services/api-manager/axiosInstance.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    "Content-type": "application/json",
+    locale: "en",
+  },
+});
+
+// Add a response interceptor
+axiosInstance.interceptors.response.use(
+  (response: AxiosResponse) => response,
+  async (failedRequest: AxiosResponse) =>
+    failedRequest.status === 412
```
Comment (hamza-mubeen-07) [inline src/services/api-manager/axiosInstance.ts]: what the reason for this logic i think we will be getting status code from backend always ?

**Review by hamza-mubeen-07 — CHANGES_REQUESTED** [review]

File: src/services/context/GlobalContext.tsx [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    ) => Promise<iApiResponse>
```
Comment (junaid1840) [inline src/services/context/GlobalContext.tsx]: Let's have a discussion on this ? 
CC: @hamza-mubeen-07 @HumnaAleem1

File: src/services/context/GlobalProvider.tsx [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  const [showLoader, setShowLoader] = useState(false);
+
+  const globalDispatch = async (
+    effect: (
+      dispatch: Dispatch<iGetCategoryOptionsActionType>
+    ) => Promise<iApiResponse>
+  ): Promise<void> => {
+    const data = await dispatch(effect);
+    console.log(data);
+  };
```
Comment (junaid1840) [inline src/services/context/GlobalProvider.tsx]: Not looking goof in terms of Typescript.

File: src/component/home/Home.tsx (hunk 5 above)
Comment (MuhammadMuneebZafar) [reply src/component/home/Home.tsx]: It's just for the testing purpose but will do it anyway.

**Review by hamza-mubeen-07 — APPROVED** [review]

**Pull request merged**